chalk-solve/src/coherence.rs

```diff
@@ -106,18 +116,24 @@ where
 
     // Build the forest of specialization relationships.
     fn build_specialization_forest(&self) -> Result<Graph<ImplId<I>, ()>, CoherenceError<I>> {
-        // The forest is returned as a graph but built as a GraphMap; this is
-        // so that we never add multiple nodes with the same ItemId.
-        let mut forest = DiGraphMap::new();
+        let mut forest = DiGraph::new();
+
+        let node_impls: Vec<ImplId<_>> = forest.raw_nodes().iter().map(|x| x.weight).collect();
 
         // Find all specializations (implemented in coherence/solve)
         // Record them in the forest by adding an edge from the less special
         // to the more special.
         self.visit_specializations_of_trait(|less_special, more_special| {
-            forest.add_edge(less_special, more_special, ());
+            // Check so that we never add multiple nodes with the same ImplId.
+            if !node_impls.contains(&less_special) && !node_impls.contains(&more_special) {
```

Comment:
I'm not sure this covers all the relevant cases.

Reply:
See https://github.com/rust-lang/chalk/pull/740/files#diff-ba814a8e084488bce444bd626693c6c99786b7be83fc1f0f05c51853e0fb96deL63-R74